src/target/riscv/riscv-013.c

```diff
@@ -4341,77 +4326,127 @@ static int read_memory_progbuf_inner_one(struct target *target,
 /**
  * Read the requested memory, silently handling memory access errors.
  */
-static mem_access_result_t
-read_memory_progbuf(struct target *target, const riscv_mem_access_args_t args)
+static mem_access_result_t read_memory_progbuf(struct target *target,
+		const riscv_mem_access_args_t args, bool mprven)
 {
 	assert(riscv_mem_access_is_read(args));
 
-	mem_access_result_t skip_reason = mem_should_skip_progbuf(target, args);
-	if (skip_reason != MEM_ACCESS_OK)
-		return skip_reason;
-
-	LOG_TARGET_DEBUG(target, "reading %" PRIu32 " elements of %" PRIu32
-			" bytes from 0x%" TARGET_PRIxADDR, args.count, args.size, args.address);
-
-	if (dm013_select_target(target) != ERROR_OK)
-		return MEM_ACCESS_SKIPPED_TARGET_SELECT_FAILED;
-
 	select_dmi(target);
-
 	memset(args.read_buffer, 0, args.count * args.size);
 
 	if (execute_autofence(target) != ERROR_OK)
 		return MEM_ACCESS_SKIPPED_FENCE_EXEC_FAILED;
 
+	int res = (args.count == 1) ?
+			read_memory_progbuf_inner_one(target, args, mprven) :
+			read_memory_progbuf_inner(target, args, mprven);
+
+	return res == ERROR_OK ? MEM_ACCESS_OK : MEM_ACCESS_FAILED;
+}
+
+static mem_access_result_t write_memory_progbuf(struct target *target,
+		const riscv_mem_access_args_t args, bool mprven);
+
+static mem_access_result_t
+access_memory_progbuf(struct target *target, const riscv_mem_access_args_t args)
+{
+	mem_access_result_t result = mem_should_skip_progbuf(target, args);
+	if (result != MEM_ACCESS_OK)
+		return result;
+
+	const bool is_read = riscv_mem_access_is_read(args);
+	const char *const access_type = is_read ? "read" : "write";
```

Comment:
This will result in "writeing" in case of a write. Consider assigning a gerund directly.
```suggestion
	const char *const access_type = is_read ? "reading" : "writing";
```

Reply:
Addressed